framework/stats/stats.cpp

```diff
@@ -0,0 +1,287 @@
+					pending_samples                  = continuous_samples;
+					continuous_samples.clear();
+				}
+			}
+
+			if (pending_samples.size() == 0)
+				return;
+
+			// Ensure the number of pending samples is capped at a reasonable value
+			if (pending_samples.size() > 100)
+				pending_samples.resize(100);
+
+			// Compute the number of samples to show this frame
+			size_t sample_count = static_cast<size_t>(sampling_config.speed * delta_time) * pending_samples.size();
+
+			// Clamp the number of samples
+			sample_count = std::max<size_t>(1, std::min<size_t>(sample_count, pending_samples.size()));
+
+			// Get the frame time stats (not a continuous stat)
+			StatsProvider::Counters frame_time_sample = frame_time_provider->sample(delta_time);
+
+			// Push the samples to circular buffers
+			std::for_each(pending_samples.end() - sample_count, pending_samples.end(), [this, frame_time_sample](auto &s) {
+				// Write the correct frame time into the continuous stats
+				s.insert(frame_time_sample.begin(), frame_time_sample.end());
+				// Then push the sample to the counters list
+				this->push_sample(s);
+			});
+			pending_samples.erase(pending_samples.end() - sample_count, pending_samples.end());
+
+			break;
+		}
+	}
+}
+
+void Stats::continuous_sampling_worker(std::future<void> should_terminate)
+{
+	worker_timer.tick();
+
+	for (auto &p : providers)
+		p->continuous_sample(0.0f);
+
+	while (should_terminate.wait_for(std::chrono::seconds(0)) != std::future_status::ready)
+	{
+		auto delta_time = static_cast<float>(worker_timer.tick());
+		auto interval   = std::chrono::duration_cast<std::chrono::duration<float>>(sampling_config.interval).count();
+
+		// Ensure we wait for the interval specified in config
+		if (delta_time < interval)
+		{
+			std::this_thread::sleep_for(std::chrono::duration<float>(interval - delta_time));
+			delta_time += static_cast<float>(worker_timer.tick());
+		}
+
+		// Sample counters
+		StatsProvider::Counters sample;
+		for (auto &p : providers)
+		{
+			StatsProvider::Counters s = p->continuous_sample(delta_time);
+			sample.insert(s.begin(), s.end());
+		}
+
+		// Add the new sample to the vector of continuous samples
+		{
+			std::unique_lock<std::mutex> lock(continuous_sampling_mutex);
+			continuous_samples.push_back(sample);
+		}
+	}
+}
+
+void Stats::push_sample(const StatsProvider::Counters &sample)
+{
+	for (auto &c : counters)
+	{
+		StatIndex           idx    = c.first;
+		std::vector<float> &values = c.second;
+
+		// Find the counter matching this StatIndex in the Sample
+		const auto &smp = sample.find(idx);
+		if (smp == sample.end())
+			continue;
+
+		float measurement = static_cast<float>(smp->second.result);
+
+		add_smoothed_value(values, measurement, alpha_smoothing);
+	}
+}
+
+void Stats::command_buffer_begun(CommandBuffer &cb)
+{
+	// Inform the providers
+	for (auto &p : providers)
+		p->command_buffer_begun(cb);
+}
+
+void Stats::command_buffer_ending(CommandBuffer &cb)
```

Comment:
Same goes for this, can this be `end_sampling`, or `finish` or something along these lines.

Reply:
Done